genapi/listener_reloader.go

```diff
@@ -58,9 +68,16 @@ func (lr *listenerReloader) Accept() (net.Conn, error) {
 
 // Close only closes the wrapping net.Listener, it's assumed the underlying one
 // will be subsequently closed down the chain
+// In the event of http.Server.Shutdown, Close will be called by the defer in
+// Serve and the Shutdown function. So we detect if we're already closed and
+// return an error if we already are. This seems consistant with builtin.
 func (lr *listenerReloader) Close() error {
-	close(lr.closeCh)
-	return (<-lr.lch).Close()
+	select {
```

Comment:
sooooooooo still a problem here. If spin doesn't happen to be in the select at the moment Close is called the default case here might hit instead. If instead we did close the closeCh, when you read from it here you can do:
```
l, ok := <-lr.closeCh
if !ok {
    return errors.New("already closed")
}
return l.Close()
```
and skip the select entirely